AO3-4676 Extend tests in invitations_controller

Hamham6 opened the pull request:
# Pull Request Checklist

<!-- Mark steps in the checklist as complete by changing `[ ]` to `[X]` -->
* [x] Have you read ["How to write the perfect pull request"](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)?
* [x] Have you read the [contributing guidelines](https://github.com/otwcode/otwarchive/blob/master/CONTRIBUTING.md)?
* [ ] Have you added [tests for any changed functionality](https://github.com/otwcode/otwarchive/wiki/Automated-Testing)? N/A only tests
* [x] Have you added the [Jira](https://otwarchive.atlassian.net) issue number
  as the *first* thing in your pull request title (e.g. `AO3-1234 Fix thing`)
* [x] Do you have fewer than 5 pull requests already open? If not, please wait
  until they are reviewed and merged before creating new pull requests.

## Issue

https://otwarchive.atlassian.net/browse/AO3-4676

## Purpose

Covers lines of invitations_controller that aren't covered yet.

## Credit

Hamham6, it/its

Hamham6 committed:
Update invitations_controller_spec.rb

Hamham6 committed:
Merge branch 'AO3-4676' of github.com:Hamham6/otwarchiveham into AO3-4676

Hamham6 commented:
I can't figure out how to get delete/save to be false to test that logic branch.

Hamham6 committed:
add tests for destroy and invite_friend

Hamham6 committed:
expand tests for admin roles

Hamham6 commented:
I can role back the last commit if the dif is too big.

Hamham6 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -30,23 +30,53 @@
-        get :manage, params: { user_id: user.login }
+        subject
 
         it_redirects_to_with_error(new_user_session_path, "Sorry, you don't have permission to access the page you were trying to reach. Please log in.")
```
Should the error messsages for admins without roles and without authorized roles be changed to "Sorry, only an authorized admin can access the page you were trying to reach."? Also, should the redirect be changed to root_url or is it supposed to be new_user_session_path? These differences are keeping me from being able to use the access_shared_examples.

Hamham6 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -194,39 +261,191 @@
+          put :update, params: { id: invitation.id, invitation: { invitee_email: new_email } }
+
+          it_redirects_to_with_notice(find_admin_invitations_path("invitation[token]" => invitation.token), "Invitation was successfully sent.")
+          expect(invitation.invitee_email).to eq(new_email)
```
I don't know why this doesn't work TT

Hamham6 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -194,39 +261,191 @@
+        end
+
+        it "errors if email is missing" do
+          allow_any_instance_of(Invitation).to receive(:update).and_return(false)
```
Even if the email is blank, the error and show template aren't loaded unless the invitation failed to save or the email didn't change. Or possibly I'm misunderstanding the code, but without this line this test fails because nothing renders.

Bilka2 replied:
Found a Jira issue for this behavior change: https://otwarchive.atlassian.net/browse/AO3-5903

Hamham6 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -194,39 +261,191 @@
+          fake_login_admin(admin)
+        end
+
+        context "when invitation creator is a user" do
```
I am confused over what
`if @user.is_a?(User)
      redirect_to user_invitations_path(@user)
    else
      redirect_to admin_invitations_path
    end`
(lines 81-85 of invitations_controller.rb) means. Does @\user here refer to the creator of the invitation or to who is deleting it? Can users delete invitations?

Hamham6 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -194,39 +261,191 @@
- 
-      delete :destroy, params: { id: invitation.id }
-      it_redirects_to_with_notice(admin_invitations_path, "Invitation successfully destroyed")
+    context "when logged in as an admin with no role" do
```
Admins without roles can apparently delete invitations? As can admins without authorized roles? An error pops up "Sorry, you don't have permission to access the page you were trying to reach. Please log in." and redirects to new user session path, but the invitation is still deleted? The same thing seems to be happening for users and guests.

Hamham6 committed:
Merge branch 'otwcode:master' into AO3-4676

Bilka2 commented on spec/controllers/invitations_controller_spec.rb:
```diff
@@ -194,39 +261,191 @@
+        subject
+
+        it_redirects_to_with_error(new_user_session_path, "Sorry, you don't have permission to access the page you were trying to reach. Please log in.")
+        expect(invitation).to exist
```
This check tries to call `exist?` or `exists?` on the object which I don't think does anything sensible (it's not defined for the record in Rails). Try making a `find` for the invitation instead, then the exist check will hit [a Rails method](https://guides.rubyonrails.org/active_record_querying.html#existence-of-objects).

Hamham6 committed:
Merge branch 'otwcode:master' into AO3-4676

Hamham6 committed:
use correct Policy

Because creating/deleteing/viewing/sorting a user's invitations is considered manageing a user

Hamham6 committed:
refactor shared_examples

so shared_example for authorized and unauthorized admins can be used seperately. For example if only one matchs wanted behavior and the other doesn't, the matching one's shared_example can be used by itself.

Hamham6 committed:
add invitation_controller shared_examples

for guest, user, and invitation owner access

Hamham6 committed:
use authorize for Invitations

and use shared_examples for index and manage now that behavior matchs

Hamham6 committed:
shared_examples for show

Hamham6 committed:
shared_examples for create and destroy

and use find_by to check if a invitation exists

Hamham6 committed:
shared_examples for invite_friend

Hamham6 committed:
shared_examples for update

Hamham6 committed:
fix variable names and syntax

rspec does not like it when the it_behaves_like keyword argument and variable have the same name, so I arbitrarily made them slightly different

Hamham6 committed:
fix syntax errors

Hamham6 committed:
guest and user shared_examples

Hamham6 committed:
fix syntax for spec and authorize

Hamham6 committed:
Merge branch 'otwcode:master' into AO3-4676

Hamham6 committed:
fix and improve tests

adds test for deleting invitation created by queue

Hamham6 committed:
Merge branch 'AO3-4676' of github.com:Hamham6/otwarchiveham into AO3-4676

Hamham6 committed:
test fixs

define user and invitation variables correctly /
use the right syntax for passing `shared_examples` parameters /
fix `admin_success` to `success_admin` /
move `it_behaves_like` tests after other tests, for readability

Hamham6 committed:
refactor invite_friend

test authorized roles manually

Hamham6 committed:
minor refactoring

Hamham6 committed:
expand test

check that email hasn't changed when update has nil email

Hamham6 committed:
cop

Hamham6 committed:
Merge branch 'AO3-4676'

Hamham6 committed:
Merge branch 'master' into AO3-4676
